### PR TITLE
TPM operation support for key/csr generation

### PR DESCRIPTION
**Pull request by AssemblyJohn**
- private/public key generation
- csr generation with TPM key
- builder-based key generation
- builder-based csr generation
- relevant testcases
- cmake for testcases with crypto proider

### CODE REVIEW

File: lib/evse_security/crypto/openssl/openssl_supplier.cpp [hunk 1]
```diff
@@ -53,6 +77,235 @@ static CertificateValidationError to_certificate_error(const int ec) {
...
+
+        int success;
+        if (key_info.private_key_pass.has_value()) {
+            success = PEM_write_bio_PrivateKey(key_bio.get(), evp_key.get(), EVP_aes_128_cbc(), NULL, 0, NULL,
+                                               (void*)key_info.private_key_pass.value().c_str());
+        } else {
+            success = PEM_write_bio_PrivateKey(key_bio.get(), evp_key.get(), NULL, NULL, 0, NULL, NULL);
+        }
+
+        if (false == success) {
```
Comment (hikinggrass) [inline lib/evse_security/crypto/openssl/openssl_supplier.cpp:130]: success is int, so shouldn't this check be:
```suggestion
        if (0 == success) {
```

Comment (hikinggrass) [reply lib/evse_security/crypto/openssl/openssl_supplier.cpp:130]: Although using false here is a bit more clear in meaning...

File: lib/evse_security/crypto/openssl/openssl_supplier.cpp [hunk 2]
```diff
@@ -53,6 +77,235 @@ static CertificateValidationError to_certificate_error(const int ec) {
...
+        params[0] = OSSL_PARAM_construct_uint("bits", &bits);
+        ctx = EVP_PKEY_CTX_ptr(EVP_PKEY_CTX_new_from_name(NULL, "RSA", "provider=tpm2"));
+    } else if (type == CryptoKeyType::EC_prime256v1 || type == CryptoKeyType::EC_secp384r1) {
+        char* group_256 = "P-256";
+        char* group_384 = "P-384";
+
+        if (type == CryptoKeyType::EC_prime256v1)
+            params[0] = OSSL_PARAM_construct_utf8_string("group", group_256, sizeof(group_256));
+        else
+            params[0] = OSSL_PARAM_construct_utf8_string("group", group_384, sizeof(group_384));
```
Comment (hikinggrass) [inline lib/evse_security/crypto/openssl/openssl_supplier.cpp:175]:
```suggestion
        if (type == CryptoKeyType::EC_prime256v1) {
            params[0] = OSSL_PARAM_construct_utf8_string("group", group_256, sizeof(group_256));
        } else {
            params[0] = OSSL_PARAM_construct_utf8_string("group", group_384, sizeof(group_384));
        }
```

File: lib/evse_security/evse_security.cpp [hunk 3]
```diff
@@ -574,14 +574,21 @@ std::string EvseSecurity::generate_certificate_signing_request(LeafCertificateTy
     std::string csr;
     CertificateSigningRequestInfo info;
 
+    // TODO(ioan): get this from the parameter when the interface will support it
+    bool use_tpm = false;
+
```
Comment (hikinggrass) [inline lib/evse_security/evse_security.cpp:579]: Is this still relevant?

**Review by hikinggrass — CHANGES_REQUESTED** [review]
just a few minor things

File: lib/evse_security/evse_security.cpp (hunk 3 above)
Comment (AssemblyJohn) [reply lib/evse_security/evse_security.cpp:579]: Yes, still relevant will become a parameter when we will modify the interface.

**Review by hikinggrass — APPROVED** [review]

**Comment by marcemmers** [comment]
This PR broke compatibility with OpenSSL 1.1.1 since it does not have a `provider.h`. That is new in OpenSSL 3.x. Was this intentional or are we going to bump the requirement form OpenSSL

**Pull request merged**